src/garage/torch/algos/ppo.py

```diff
@@ -0,0 +1,96 @@
+        baseline : The baseline.
+        max_path_length (int): Maximum length of a single rollout.
+        policy_lr (float): Learning rate for training policy network.
+        lr_clip_range (float): The limit on the likelihood ratio between
+            policies.
+        n_samples (int): Number of train_once calls per epoch.
+        discount (float): Discount.
+        gae_lambda (float): Lambda used for generalized advantage
+            estimation.
+        center_adv (bool): Whether to rescale the advantages
+            so that they have mean 0 and standard deviation 1.
+        positive_adv (bool): Whether to shift the advantages
+            so that they are always positive. When used in
+            conjunction with center_adv the advantages will be
+            standardized before shifting.
+        optimizer (object): The optimizer of the algorithm. Should be the
+            optimizers in torch.optim.
+        policy_ent_coeff (float): The coefficient of the policy entropy.
+            Setting it to zero would mean no entropy regularization.
+        use_softplus_entropy (bool): Whether to estimate the softmax
+            distribution of the entropy to prevent the entropy from being
+            negative.
+        stop_entropy_gradient (bool): Whether to stop the entropy gradient.
+        entropy_method (str): A string from: 'max', 'regularized',
+            'no_entropy'. The type of entropy method to use. 'max' adds the
+            dense entropy to the reward for each time step. 'regularized' adds
+            the mean entropy to the surrogate objective. See
+            https://arxiv.org/abs/1805.00909 for more details.
+    """
+
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 baseline,
+                 max_path_length=500,
+                 policy_lr=1e-2,
+                 lr_clip_range=1e-2,
+                 n_samples=1,
+                 discount=0.99,
+                 gae_lambda=1,
+                 center_adv=True,
+                 positive_adv=False,
+                 optimizer=None,
+                 policy_ent_coeff=0.0,
+                 use_softplus_entropy=False,
+                 stop_entropy_gradient=False,
+                 entropy_method='no_entropy'):
+        super().__init__(env_spec, policy, baseline, max_path_length,
+                         policy_lr, n_samples, discount, gae_lambda,
+                         center_adv, positive_adv, optimizer, policy_ent_coeff,
+                         use_softplus_entropy, stop_entropy_gradient,
+                         entropy_method)
+
+        self._constraint_clip_range = lr_clip_range
+        self._old_policy_state_dict = self.policy.state_dict()
+
+    def _compute_objective(self, advantages, obs, actions, rewards):
+        # Compute constraint
+        current_policy_state_dict = self.policy.state_dict()
+        self.policy.load_state_dict(self._old_policy_state_dict)
```

Comment:
we don't need to load the old policy params every time, especially considering we will be doing multiple gradient steps from the old policy, before updating the old policy. I am not sure if this is correct. @krzentner. 

I think that it is best for us to keep around the old policy as an object, by using copy.deepcopy, and only updating the old policy after the we are at the last gradient step for the iteration.

@yonghyuc

Reply:
I also thought about using deepcopy method but comments says using deepcopy method could raise an error in certain PyTorch version

https://discuss.pytorch.org/t/are-there-any-recommended-methods-to-clone-a-model/483/15
https://github.com/pytorch/pytorch/issues/415#issuecomment-271226740

Reply:
resolved offline.